### PR TITLE
Handle all memory access via MemoryManager

### PR DESCRIPTION
**Pull request by sporksmith**
* Move common types from descriptor/mod.rs to syscall_types.rs
* Remove cursor functionality from TypedPluginPtr, since it    
made the common cases unnecessarily complex.
* Move memory access helpers from process.rs to memory_manager.rs    
* Split MemoryMapper out of MemoryManager    
* Create memory reader and writer traits, and implementations that    
use either process_vm_readv & process_vm_writev, or direct
memory mapped access.
* MemoryManager is now created early and should be used for all access.    
The MemoryMapper is still created lazily.
* Change process_* memory helpers to use the MemoryManager API instead    
of thread APIs.
* Remove thread_* memory helpers since they're no longer needed.    
* Update several syscall handlers to avoid non-exclusive mutable access    
to the Memory
* Remap whole stack at once, instead of doing it lazily

Incidental changes along the way:
* ByteQueue: bugfix for large aligned writes
* test_clone: avoid tripping stack smash detection
* Switch from gnu99 to gnu11 C standard in src/main
* Switch stacks when making syscalls via IPC

### CODE REVIEW

**Comment by sporksmith** [comment]
Sorry for the large PR. I broke out some of the independent changes into smaller commits, but it ended up a bit "[draw the rest of the owl](https://i.kym-cdn.com/photos/images/original/000/572/078/d6d.jpg)"

File: src/main/host/syscall/poll.c [hunk 1]
```diff
@@ -220,13 +220,18 @@ SysCallReturn syscallhandler_ppoll(SysCallHandler* sys, const SysCallArgs* args)
...
     if (!ts_timeout_ptr.val) {
         // NULL timespec is valid and means infinite timeout
         ts_timeout = NULL;
     } else {
-        ts_timeout = process_getReadablePtr(sys->process, ts_timeout_ptr, sizeof(*ts_timeout));
+        if (process_readPtr(
+                sys->process, &ts_timeout_val, ts_timeout_ptr, sizeof(ts_timeout_val)) != 0) {
+            return (SysCallReturn){.state = SYSCALL_DONE, .retval.as_i64 = -EFAULT};
+        }
+        ts_timeout = &ts_timeout_val;
```
Comment (stevenengler) [inline src/main/host/syscall/poll.c:234]: Can we remove the second timespec `ts_timeout` here?

**Review by stevenengler — CHANGES_REQUESTED** [review]
Looks good. I marked as requesting changes only because of the "opaque" types comment in cbindgen.toml.

File: src/main/CMakeLists.txt [hunk 2]
```diff
@@ -19,7 +19,7 @@ add_cflags(-fPIC)
 #add_cflags(-Wno-unused-command-line-argument)
 #add_cflags(-Wno-unknown-warning-option)
 #add_cflags(-Wno-visibility)
-add_cflags(-std=gnu99)
+add_cflags(-std=gnu11)
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/CMakeLists.txt]: Done

File: src/main/bindings/c/cbindgen.toml [hunk 3]
```diff
@@ -26,5 +26,7 @@ rename_variants = "ScreamingSnakeCase"
 # Avoid exporting C types back through again.
 exclude = ["PluginPtr", "SysCallReg", "SysCallArgs"]
 include = ["InterposeMethod", "QDiscMode"]
-# Generate all item types, excluding opaque and enum types
-item_types = ["constants", "globals", "structs", "unions", "typedefs", "functions"]
+# Generate all item types, excluding enum types.
+# We have to export opaque again here for this header to be self-consistent,
+# even though they're already exported in bindings-opaque.h.
+item_types = ["constants", "globals", "structs", "unions", "typedefs", "opaque", "functions"]
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/bindings/c/cbindgen.toml]: Clarified the comment

File: src/main/host/syscall/poll.c (hunk 1 above)
Comment (sporksmith) [reply src/main/host/syscall/poll.c]: Sort of - I got rid of the local pointer and just did the "pointer selection" in the call to `_syscallhandler_pollHelper`. Commented and rearranged a bit to hopefully clarify what's going on here.

File: src/main/bindings/rust/CMakeLists.txt [hunk 4]
```diff
@@ -24,11 +24,8 @@ add_custom_command(OUTPUT wrapper.rs
         --whitelist-function "descriptor_unref"
         --whitelist-function "descriptor_setHandle"
         --whitelist-function "process_.*CompatDescriptor"
-        --whitelist-function "process_get.*Ptr"
         --whitelist-function "process_parseArgStr.*"
-        --whitelist-function "process_read.*Ptr"
-        --whitelist-function "process_write.*Ptr"
-        --whitelist-function "process_flushPtrs"
+        --whitelist-function "process_getMemoryManager"
         --whitelist-function "process_getMemoryManager"
```
[the remark this answers is not in the document]
Comment (sporksmith) [reply src/main/bindings/rust/CMakeLists.txt]: Fixed

**Review by stevenengler — APPROVED** [review]

**Pull request merged**